src/client_side.cc

```diff
@@ -2556,33 +2561,35 @@ httpsEstablish(ConnStateData *connState, const Security::ContextPointer &ctx)
 }
 
 #if USE_OPENSSL
-/**
- * A callback function to use with the ACLFilledChecklist callback.
- */
-static void
-httpsSslBumpAccessCheckDone(Acl::Answer answer, void *data)
+void
+HttpsSslBumpStep1AccessCheckDone(Acl::Answer answer, void *data)
 {
-    ConnStateData *connState = (ConnStateData *) data;
+    auto connState = static_cast<ConnStateData *>(data);
+    connState->httpsSslBumpStep1AccessCheckDone(answer);
+}
 
+void
+ConnStateData::httpsSslBumpStep1AccessCheckDone(const Acl::Answer answer)
+{
     // if the connection is closed or closing, just return.
-    if (!connState->isOpen())
+    if (!isOpen())
         return;
 
     if (answer.allowed()) {
-        debugs(33, 2, "sslBump action " << Ssl::bumpMode(answer.kind) << "needed for " << connState->clientConnection);
-        connState->sslBumpMode = static_cast<Ssl::BumpMode>(answer.kind);
+        debugs(33, 2, "sslBump action " << Ssl::bumpMode(answer.kind) << "needed for " << clientConnection);
+        sslBumpMode = static_cast<Ssl::BumpMode>(answer.kind);
```

Comment:
@chtsanti, this slightly modified official code sets sslBumpMode. I have two or three questions about this:

1. The code does _not_ call sslBumpNeed() so it does not set sslBumpNeed_. Should it? When (i.e. under what conditions) should we set that logging field?

    I suspect that the lack of that sslBumpNeed() call or its equivalent breaks ClientHttpRequest::doCallouts() part that is supposed to bump the connection on _error_ (see the code above the `ClientHttpRequest::doCallouts+sslBumpNeeded` string): For example, I have a customer log showing that, after visiting this code, and then encountering a request error (a REQMOD service blocks the fake CONNECT request), Squid uses the `ClientHttpRequest::doCallouts-sslBumpNeeded` branch instead (`-` instead of `+`).

2. The code does _not_ set `http->al->ssl.bumpMode` either. Should it? When (i.e. under what conditions) should we set that logging field?

3. If your answers above do not make that clear already, what is the key difference between `sslBumpMode` and `sslBumpNeed_`? The documentation quoted below implies that sslBumpNeed_ is about the _future_ bumping action but this httpsSslBumpStep1AccessCheckDone() code seems to be using sslBumpMode for that future action instead.

```
src/client_side_request.h: /// whether (and how) the request needs to be bumped
src/client_side_request.h: Ssl::BumpMode sslBumpNeed_;

src/client_side.h: Ssl::BumpMode sslBumpMode; ///< ssl_bump decision

src/AccessLogEntry.h: int bumpMode; ///< whether and how the request was SslBumped
```

Reply:
> @chtsanti, this slightly modified official code sets sslBumpMode. I have two or three questions about this:
> 
>     1. The code does _not_ call sslBumpNeed() so it does not set sslBumpNeed_. Should it? When (i.e. under what conditions) should we set that logging field?
>        I suspect that the lack of that sslBumpNeed() call or its equivalent breaks ClientHttpRequest::doCallouts() part that is supposed to bump the connection on _error_ (see the code above the `ClientHttpRequest::doCallouts+sslBumpNeeded` string): For example, I have a customer log showing that, after visiting this code, and then encountering a request error (a REQMOD service blocks the fake CONNECT request), Squid uses the `ClientHttpRequest::doCallouts-sslBumpNeeded` branch instead (`-` instead of `+`).
>

My opinion is that the ClientHttpRequest::sslBumpNeed_ and ClientHttpRequest::sslBumpNeeded() should be private to ClientHttpRequest class. Should not be set and used by external users.
 
If your observation is correct then this is a case it is not covered well inside ClientRequestContext::sslBumpAccessCheck() method. Look at the ``` "if (bumpMode != Ssl::BumpEnd) {" ``` block and specially inside the ```if (!http->getConn()->switchedToHttps() &&``` block inside this method.

Maybe we can change this policy and set the sslBumpNeed_ from client_side.cc code however, if it is easier. This part of code has survived from many transformation, before we start using fakeAConnectRequest mechanism which simplified a lot the related code.
 
>     2. The code does _not_ set `http->al->ssl.bumpMode` either. Should it? When (i.e. under what conditions) should we set that logging field?

This is set inside ClientRequestContext::sslBumpAccessCheck() too.

> 
>     3. If your answers above do not make that clear already, what is the key difference between `sslBumpMode` and `sslBumpNeed_`? The documentation quoted below implies that sslBumpNeed_ is about the _future_ bumping action but this httpsSslBumpStep1AccessCheckDone() code seems to be using sslBumpMode for that future action instead.

The AccessLogEntry::ssl::BumpMode and ConnStateData::sslBumpMode are set for the client connection, they are set and used by all requests tunneled through the bumped connection.
The sslBumpNeed_ should set only for the CONNECT (or fake CONNECT) request.

> 
> 
> ```
> src/client_side_request.h: /// whether (and how) the request needs to be bumped
> src/client_side_request.h: Ssl::BumpMode sslBumpNeed_;
> 
> src/client_side.h: Ssl::BumpMode sslBumpMode; ///< ssl_bump decision
> 
> src/AccessLogEntry.h: int bumpMode; ///< whether and how the request was SslBumped
> ```

Reply:
@eduard-bagdasaryan, I am not sure my interpretation is accurate, but if I am interpreting the implications behind @chtsanti [comment](https://github.com/measurement-factory/squid/pull/92#discussion_r510483935) correctly, ClientRequestContext::sslBumpAccessCheck() needs more work for this PR code to work correctly as far as bump-to-deliver-an-error functionality is concerned. Here are the corresponding cache.log lines, showing that sslBumpAccessCheck() was called before doCallouts() incorrectly decides not to bump the connection to deliver the error (the `doCallouts-sslBumpNeeded` code path):

```
2020/10/21 08:24:27.151 kid1| 33,2| client_side.cc(2579) httpsSslBumpStep1AccessCheckDone: sslBump action peekneeded...
2020/10/21 08:24:27.239 kid1| 93,3| XactionRep.cc(447) blockVirgin:  [M3 !V A0 job24]
2020/10/21 08:24:27.242 kid1| 85,5| client_side_request.cc(1421) sslBumpAccessCheck: SslBump already decided (6), ignoring ssl_bump...
2020/10/21 08:24:27.251 kid1| 20,3| store.cc(466) unlock: ClientHttpRequest::doCallouts-sslBumpNeeded...
```

Are you comfortable fixing this code or is it better to bounce this issue to Christos so that he can address this SslBump-specific problem?

Reply:
I can try to fix it myself, but I need some time (an hour or two) to restore the context and figure it out.

Reply:
Your call. Just let me and Christos know your decision.

Reply:
> ClientRequestContext::sslBumpAccessCheck() needs more work for this PR code

Probably we need to fix this method (maybe addressing the comment there about bug 4340), but it needs further investigation. However, (after examining the PR code) I think that the PR changes are unrelated - the same problem must exist in master as well.

> it better to bounce this issue to Christos 

Yes, I would ask @chtsanti to look at this problem (it is the area he is more familiar with, I guess:wink:)

BTW, I have a question about ClientRequestContext::sslBumpAccessCheck(). This method is called from doCallouts(), but at the moment step1 checks have completed already (in master they are initiated by postHttpsAccept()/httpsSslBumpAccessCheckDone()). So if step1 checks succeeded, connState->sslBumpMode will get a value, and sslBumpAccessCheck() will assume that 'the bumping decision has been made', exiting without performing its checks (I wonder whether these checks in sslBumpAccessCheck() are reachable in principle). @chtsanti , do you think this is the correct behavior? Is there a scenario when sslBumpAccessCheck() starts its checks?

Reply:
>  do you think this is the correct behavior? Is there a scenario when sslBumpAccessCheck() starts its checks?

Yes. This is is the correct behavior.